pkg/healthcheck/healthcheck_multicluster.go

```diff
@@ -491,3 +507,54 @@ func (hc *HealthChecker) checkRemoteClusterGatewaysHealth(ctx context.Context) e
 	}
 	return &SkipError{Reason: "not checking muticluster"}
 }
+
+func (hc *HealthChecker) checkIfMirroredServicesHaveEndpoints() error {
+	if hc.Options.ShouldCheckMulticluster {
+
+		var servicesWithNoEndpoints []string
+		mirroredServices, err := hc.kubeAPI.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{LabelSelector: k8s.MirroredResourceLabel})
+		if err != nil {
+			return err
+		}
+
+		for _, svc := range mirroredServices.Items {
+			// Check if there is a relevant end-point
+			_, err := hc.kubeAPI.CoreV1().Endpoints(svc.Namespace).Get(svc.Name, metav1.GetOptions{})
+			if err != nil {
+				servicesWithNoEndpoints = append(servicesWithNoEndpoints, fmt.Sprintf("%s.%s of cluster: [%s], gateway: [%s/%s]", svc.Name, svc.Namespace, svc.Labels[k8s.RemoteClusterNameLabel], svc.Labels[k8s.RemoteGatewayNsLabel], svc.Labels[k8s.RemoteGatewayNameLabel]))
+			}
+		}
+
+		if len(servicesWithNoEndpoints) > 0 {
+			return fmt.Errorf("Some mirrored services do not have endpoints:\n\t%s", strings.Join(servicesWithNoEndpoints, "\n\t"))
+		}
+		return nil
+	}
+	return &SkipError{Reason: "not checking muticluster"}
+}
+func (hc *HealthChecker) checkifGatewaysExistforExportedServices() error {
+	if hc.Options.ShouldCheckMulticluster {
```

Comment:
So what are you checking here, the target or the source cluster? I am not sure this is correct because: 
- gateway mirrors will be created in the namnespace where the mirror controller lives not in the namespace where they live on the target cluster
- not sure that using the `GatewayNameAnnotation` will work because we append the name of the cluster at the end (for gateway mirrors)

Reply:
This check is done on the target cluster, to see if there is a gateway service associated with each exported service. We do this by getting the exported service and reading the `gateway-name` & `gateway-namespace` annotation. I did not actually see the cluster-name being appended here (maybe it is on hte source cluster).

Example exported service

```yaml
 kubectl --context westus -n emojivoto get svc voting-svc -oyaml                   apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"mirror.linkerd.io/gateway-name":"linkerd-gateway","mirror.linkerd.io/gateway-ns":"linkerd-multicluster"},"creationTimestamp":"2020-06-03T12:06:51Z","name":"voting-svc","namespace":"emojivoto","resourceVersion":"483451","selfLink":"/api/v1/namespaces/emojivoto/services/voting-svc","uid":"0be02047-63a5-4f12-bb5f-7082a01b577f"},"spec":{"clusterIP":"10.0.247.27","ports":[{"name":"grpc","port":8080,"protocol":"TCP","targetPort":8080},{"name":"prom","port":8801,"protocol":"TCP","targetPort":8801}],"selector":{"app":"voting-svc"},"sessionAffinity":"None","type":"ClusterIP"},"status":{"loadBalancer":{}}}
    mirror.linkerd.io/gateway-name: linkerd-gateway
    mirror.linkerd.io/gateway-ns: linkerd-multicluster
  creationTimestamp: "2020-06-03T12:06:51Z"
  name: voting-svc
```

Reply:
Ahh I get it now. However in my mind this is a bit confusing. I think the whole multicluster checks need to be relevant only in the context on one cluster target or source. Now we have checks that are checking both the soruce and the target cluster. In my mind we can simplify that by sayin that we only run checks on the source cluster and we check every target cluster that we have config for for the things we want to check. Does that make sense? Similar to how we go about the trust root anchors, we can iterate through all target clusters we are paired with, grab the config, instantiate a k8s api and check whether it is all good on the other side. I think we are fine because we have permission to look at services on target clusters. Does that make sense or you think its  not such a good idea? @grampelberg WDYT ?

Reply:
That totally makes sense. My only concern being, If I own the target cluster, I can't directly check if there are any configured services?

Apart from that I'm okay with the proposed change, and let @grampelberg  decide! :)

Reply:
As discussed today, I think doing any checks on the remote cluster is overkill.  However, we should check the health of the local cluster as a source (check that mirror services and mirror gateways are valid) AND as a target (check that exported services and gateways are valid).

Reply:
Hmmmm, :thinking: I remember us concluding that we should do both so that users on the source cluster can also understand directly if there are any problems with the target clusters that they access (as its effect the client services in their cluster).

I updated the code to perform these checks on both sides, but happy to revisit if this was not intended.